File: 1488-avoid-flood-in-the-city/1488-avoid-flood-in-the-city.java
class Solution {
    public int[] avoidFlood(int[] rains) {
        int n= rains.length;
        int[] ans= new int[n];
        Map<Integer, Integer> last_rain= new HashMap<>();
        int def= 5;
        TreeSet<Integer> dry_days= new TreeSet<>();
        for(int i=0; i<n; i++){
            int lake= rains[i];
            if(lake==0){
                dry_days.add(i);
                ans[i]= def;
            }else{
                ans[i]= -1;
                if(last_rain.containsKey(lake)){
                    Integer day= dry_days.higher(last_rain.get(lake));
                    //next dry day of last_rain.get(lake)
                    //[2 , 3, 4, 5, 9, 10]
                    //6
                    if(day==null) return new int[0];
                    ans[day]= lake;
                    dry_days.remove(day);
                }
                last_rain.put(lake, i);
            }
        }
        return ans;
    }
}

File: 1488-avoid-flood-in-the-city/1488-avoid-flood-in-the-city.cpp
class Solution {
public:
    vector<int> avoidFlood(vector<int>& rains) {

        int n = rains.size();
        vector<int> ans(n, 0);

        unordered_map<int,int> mp;

        

       set<int> index;
        
        for (int i = 0; i < n; i++) {

            if(rains[i]==0){
                index.insert(i);
                continue;
            }

            if (mp.find(rains[i]) != mp.end()) {
                if (index.size() == 0) {
                    return {};
                }

                else {
                   
                   auto it = index.lower_bound(mp[rains[i]]);

                   if(it==index.end()){
                    return {};
                   }

                   ans[*it]=rains[i];

                   index.erase(it);
                 

                   
                }
            }

            ans[i]=-1;
            mp[rains[i]]=i;
        }


        for(int i=0;i<n;i++){
            if(ans[i]==0){
                ans[i]=1;
            }
        }

        return ans;
    }
};